mining.go

```diff
@@ -2055,3 +2035,231 @@ func (g *BlkTmplGenerator) UpdateBlockTime(header *wire.BlockHeader) error {
+	g.templatePoolMtx.Unlock()
+}
+
+// pruneOldBlockTemplates prunes all old block templates from the template pool.
+func (g *BgBlkTmplGenerator) pruneOldBlockTemplates(bestHeight int64) {
+	g.templatePoolMtx.Lock()
+	for rootHash, block := range g.templatePool {
+		height := int64(block.Header.Height)
+		if height < bestHeight-1 {
+			delete(g.templatePool, rootHash)
+		}
+	}
+	g.templatePoolMtx.Unlock()
+}
+
+// UpdateParentTemplate updates parent template with the the current template.
+// A new block template is generated immediately if the block height is below
+// stake validation height. This should be called when a new block has been
+// accepted.
+func (g *BgBlkTmplGenerator) UpdateParentTemplate() {
+	stakeValidationHeight := g.tg.chainParams.StakeValidationHeight
+	g.templateMtx.Lock()
+	var height int64
+
+	if g.currentTemplate != nil {
+		height = g.currentTemplate.Height
+
+		// Set the parent template if the chain is matured to stake
+		// validation height.
+		if height >= stakeValidationHeight-2 {
+			g.parentTemplate, g.currentTemplate = g.currentTemplate, nil
+			minrLog.Debugf("Parent template updated to height: %v",
+				g.parentTemplate.Height)
+		}
+
+		// Prune invalidated block templates.
+		g.pruneOldBlockTemplates(height)
+	}
+	g.templateMtx.Unlock()
+
+	// Regenerate a new block template based off the newly
+	// connected block if the current block height is below stake
+	// validation height.
+	if height < stakeValidationHeight-2 {
+		g.regenTemplate()
+	}
+}
+
+// regenTemplateHandler regenerates the block template if there are enough
+// votes for the current chain tip.
+func (g *BgBlkTmplGenerator) regenTemplateHandler(ctx context.Context) {
+	for {
+		select {
+		case <-g.regenChan:
+			// Get the number of votes in for the current chain tip.
+			votes := len(g.tg.txSource.
+				VoteHashesForBlock(&g.tg.chain.BestSnapshot().Hash))
+
+			// Regenerate the block template only if the number of votes for the
+			// current chain tip is greater than or equal the minimum required
+			// for network.
+			if votes >= int((g.tg.chainParams.TicketsPerBlock/2)+1) {
+				g.regenTemplate()
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// RegenTemplate signals block template regeneration.
+func (g *BgBlkTmplGenerator) RegenTemplate() {
+	select {
+	case g.regenChan <- struct{}{}:
+	default:
+		// Non-blocking send fallthrough.
+	}
+}
+
+// generator generates new templates based on mempool activity for vote
+// and non-vote transactions and the time elapsed since last template
+// regeneration. This must be run as a goroutine.
+func (g *BgBlkTmplGenerator) generator(ctx context.Context) {
+	ticker := time.NewTicker(time.Millisecond * 500)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			lastRegen := atomic.LoadInt64(&g.lastRegen)
+			lastUpdated := g.tg.txSource.LastUpdated().Unix()
+			now := time.Now().Unix()
+			sinceLastRegen := now - lastRegen
+
+			if !g.started {
```

Comment:
This is only used here in this goroutine.  Prefer a local over a variable in the state.

Also, no need to set the `sinceLastRegen`.  Just set `lastRegen` and `started` and then `continue`.